### ISSUE TITLE
Latest dev builds: first performance drop, now NoSuitableGraphicsDeviceException?

### ISSUE THREAD

**Issue by yyr57695** [MonoGame/MonoGame#5449]
I am running MonoGame on Windows with Visual Studio 2015 Community.  I recently ported an XNA project that utilizes 3D graphics.  Under XNA it was using the HiDef profile.

Using MonoGame 3.5.1 the performance was indistinguishable from XNA, but there was a bug using the Xbox One controller.  If it was plugged in when the game was launched, it didn't work until it was unplugged, then reconnected.

I decided to try using the latest dev build to see if that bug was fixed.  I grabbed it as of a few weeks ago, version 3.6.0.1365.  At first, it didn't run at all...there were some minor changes I had to make, but the biggest issue was an error telling me I had to use 16-bit indices rather than 32-bit indices.  Huh?  I ultimately realized that it was trying to force the Reach profile, and fixed it by adding the following line in my constructor:

graphics.GraphicsProfile = GraphicsProfile.HiDef;

where "graphics" is my GraphicsDeviceManager.  This got the game running, and the controller detection bug had indeed been fixed in the new MonoGame build, but the framerate was approximately half of what it had been before.  The game looked identical, but the performance had dropped like a rock.

I was hoping that this was just a temporary bug.  I tried again with version 3.6.0.1425 just a few days ago (which is still the current version as of 1/29/2017).  But now the game won't start at all.  It builds OK, but crashes during initialization with a Microsoft.Xna.Framework.Graphics.NoSuitableGraphicsDeviceException.

I am using a 2010 MacBook sporting an integrated GeForce 320M GPU.  Yeah, an antique by today's standards, but it always ran my XNA stuff quite easily.  

Were the hardware requirements just raised?  If so, why?  If not, what could be causing this?

If there's no easy fix for this (and the performance issues), is it possible for me to Frankenstein the new Input code into MonoGame 3.5.1 and use that instead, for its superior performance?

Thanks very much for your assistance.

**Comment by Jjagg** [comment]
> biggest issue was an error telling me I had to use 16-bit indices rather than 32-bit indices. Huh? I ultimately realized that it was trying to force the Reach profile

Changes have been made ik 3.6 in an effort to actually start enforcing the GraphicsProfile. Two issues here:

1. There are some cases were no decent exception messages are being thrown because of missing functionality due to GraphicsProfile limitations. (e.g. #5108, which a lot of people have run in to).
2. It seems profiles don't map correctly to XNA's profiles. If your game ran in Reach in XNA it should run in Reach in MonoGame without issues.

It's good to know about this particular issue, so thanks for reporting it.

> the framerate was approximately half of what it had been before

Woah. That shouldn't happen. Though you should note that half the framerate isn't actually half the performance. Generally speaking, framerate is actually a pretty bad performance metric :p For talking about relative performance you should invert framerate and talk about *frame time*. 
Any chance you could profile this?  What kind of game is it?

> but crashes during initialization with a Microsoft.Xna.Framework.Graphics.NoSuitableGraphicsDeviceException.

Does your graphics card support the HiDef profile in XNA? A GraphicsProfile requires some minimal supported level of specific functionality and that is also something that is being enforced in 3.6. The recently added check that sees if HiDef is supported is too strict however. See Tom's comment here: https://github.com/MonoGame/MonoGame/pull/5312#issuecomment-273285950 

Also related here; there are plans for rewriting GraphicsCapabilities completely to expand beyond reach and hidef #1941.

**Comment by yyr57695** [comment]
Thank you muchly for your prompt and detailed reply!

I glanced at #5108 and that's definitely not in play on my end.  While running in XNA, this particular game was using the HiDef profile from the start; I never did target Reach with it.  

Just in case it will help, I'll elaborate on what happened under 3.6.0.1365:

The game launched okay, but as soon as it began the routines to display 3D graphics, and attempted to create a new IndexBuffer, this happened:

```
An unhandled exception of type 'System.NotSupportedException' occurred in MonoGame.Framework.dll

Additional information: The profile does not support an elementSize of IndexElementSize.ThirtyTwoBits; use IndexElementSize.SixteenBits or a type that has a size of two bytes.
```

This is how I discovered that the issue with my indices was related to the graphics profile; when I searched this error message, it eventually led me to the realization that if I'd received this error message from XNA, it would be referring to a limitation of the Reach profile.  That was when I decided to add 

`graphics.GraphicsProfile = GraphicsProfile.HiDef;`

to my code, which got the game running.  But as I mentioned, performance had significantly dropped.

I don't have any experience with profiling...nor do I know whether much of that is possible in the free Community version of Visual Studio.  I can say this much: I would imagine that it's not very well-optimized.  It's my first and only attempt at 3D graphics: a Snake game that supports up to 4 simultaneous players.  But it did perform well enough under XNA and MonoGame 3.5.  I can say for certain that this system's graphics hardware supports XNA HiDef, as I've been running the game on it for years.

For now...I found the page with all of the old builds...I'll see if I can figure out at what version the performance dropped, so maybe we can figure out what else happened.

Hope this info helps in some way.

**Comment by nkast** [comment]
Under XNA you can set the required profile of the project from the project properties. You probably had that set to HiDef from the beginning.

Under MG3.5 the Profile was set to HiDef by default if the hardware support it.
The problem with that behavior was that most users didn't set it to Reach and thous limit their audience  even if they didn't need HiDef.

Under MG3.6 the Profile is set to Reach by default. If you require HiDef you can set it to HiDef just like you did. MG3.6 now has `GraphicsAdapter.IsProfileSupported()` which you can use to either present a custom error message, or fall back to Reach profile. See #5312.

**Comment by Jjagg** [comment]
> I glanced at #5108 and that's definitely not in play on my end. 

Oh no , I referred to it as an example of how we do not throw correct exceptions in all cases because of insufficient graphicsprofile. But it seems that's not a problem here :)

> Under XNA you can set the required profile of the project from the project properties. You probably had that set to HiDef from the beginning.

Okay, I didn't know that. That clarfies things for me.

> Or is it not possible to download a dev build that's more than a few weeks old?

It is not. And it would be really slow to find the build that caused it this way, so thanks for wantong to try, but you should either build from source or just leave it to us :)

> Hope this info helps in some way.

It does, thanks for wanting to help out :)

I have two more questions:

1. Are you setting the. GraphicsProfile in your game constructor (I'm guessing yes, but if not you should)
2. Do you have both a dedicated and integrated. GPU in your device? Maybe the dedicated one was picked before and it's now the integrated card?

**Comment by yyr57695** [comment]
Too bad about not being able to grab the old builds.  But I suppose moving forward is more important than going back...

Yes, the line setting the GraphicsProfile is early on in the constructor for the Game class.

I do not have multiple GPUs in this MacBook.  It's the Mid-2010 model, which exclusively utilizes the GeForce 320M.

I popped open the latest XNA version of this project, and can confirm that I have always used the HiDef profile with it.

If you have any other questions or if there's anything else I can do to assist please let me know.  For the moment, I think I might either suspend this project temporarily, or continue using good old XNA, and then port it back to MonoGame again once it's feature-complete.

**Comment by zigzag312** [comment]
Hi guys! (I just started porting some XNA stuff to MonoGame)

I also stumbled upon NoSuitableGraphicsDeviceException using latest develop branch. The exception occurs if multisampling is enabled. When PreferMultiSampling is true, MultiSampleCount is set to 32 inside GraphicsDeviceManager.PreparePresentationParameters method. While debugging this method I noticed that GraphicsDevice is null, and hardcoded value of 32 samples is used. When I changed the hardcoded value to 4, it didn't crash (I don't know if multisampling actually worked as I wasn't drawing anything in my test).

yyr57695 do you also have multisampling enabled? Multisampling could also explain performance drop.

**Comment by MichaelDePiazzi** [comment]
> The exception occurs if multisampling is enabled.

I believe this might be fixed by PR #5446

**Comment by zigzag312** [comment]
I tested PR #5446. It fixes the crash, but it does so by completely disabling multisampling.

I think the fix that PR #5446 tries to implement should be moved to GetMultiSamplingQuality method. Because according to DX11 documentation it should never return -1:

> The maximum number of quality levels is defined by D3D11_MAX_MULTISAMPLE_SAMPLE_COUNT in D3D11.h. If this method returns 0, the format and sample count combination is not supported for the installed adapter.
Furthermore, the definition of a quality level is up to each hardware vendor to define, however no facility is provided by Direct3D to help discover this information. [source](https://msdn.microsoft.com/en-us/library/windows/desktop/ff476499(v=vs.85).aspx)

`private int GetMultiSamplingQuality(Format format, int multiSampleCount)
        {
            return _d3dDevice.CheckMultisampleQualityLevels(format, multiSampleCount) - 1;
        }`

-1 is is there because:

> The valid range is **between zero and one less than** the level returned by ID3D10Device::CheckMultisampleQualityLevels for Direct3D 10 or ID3D11Device::CheckMultisampleQualityLevels for Direct3D 11.  [source](https://msdn.microsoft.com/en-us/library/windows/desktop/bb173072(v=vs.85).aspx)

...but we should also clamp the value so it doesn't go below zero. 

To find valid multisampling level we could call GetMultiSamplingQuality method  with multiSampleCount of 32, 16, 8, 4, 2 until we get quality that is higher than 0 or we reach end of enumeration. Method name that implements this could be TryGetMultiSamplingQuality or something like that. (I am _guessing_ value needs to be power of two....need to verify that). 

I can try to create new pull request, unless there is a better way to find valid multisampling level?

**Comment by MichaelDePiazzi** [comment]
@zigzag312 - I'm not the best person to ask about this as I've never really messed with multisampling. I just noticed PR #5446 and thought it might help with the crash you were having. @KonajuGames might be in a better position to comment on this than me (he created the PR)...

@KonajuGames - Any thoughts on @zigzag312's comments?

**Comment by Jjagg** [comment]
> To find valid multisampling level we could call GetMultiSamplingQuality method with multiSampleCount of 32, 16, 8, 4, 2 until we get quality that is higher than 0 or we reach end of enumeration. Method name that implements this could be TryGetMultiSamplingQuality or something like that. (I am guessing value needs to be power of two....need to verify that).

We already do that. The maximum supported multisample count is stored in the GraphicsCapabilities when creating the device. https://github.com/MonoGame/MonoGame/blob/develop/MonoGame.Framework/Graphics/GraphicsCapabilities.DirectX.cs#L35
But there are issues with the current implementation.

The ms count is clamped when a new GraphicsDevice is created in GraphicsDeviceManager. https://github.com/MonoGame/MonoGame/blob/develop/MonoGame.Framework/GraphicsDeviceManager.cs#L214 
However that is not the right place to do the clamping because users can directly call GraphicsDevice.Reset with some custom parameters. Another issue is we don't know the max multisample level when creating the first device and even get it wrong when switching to a different device... We should instead do the clamping in GraphicsDevice at https://github.com/MonoGame/MonoGame/blob/develop/MonoGame.Framework/Graphics/GraphicsDevice.DirectX.cs#L386 for UWP and at https://github.com/MonoGame/MonoGame/blob/develop/MonoGame.Framework/Graphics/GraphicsDevice.DirectX.cs#L724 for regular DX. At that point the MaxMultiSampleCount for the GraphicsDevice has already been determined.

**Comment by zigzag312** [comment]
I see. I will try to come up with a solution. GraphicsCapabilities is updated too late here
https://github.com/MonoGame/MonoGame/blob/develop/MonoGame.Framework/Graphics/GraphicsDevice.cs#L272
when maximum supported multisample count is already needed at method call that is made one line before.

**Comment by zigzag312** [comment]
I have created a fix for DirectX version that seems to be working, but I would like to discuss some changes I made that affect other windows platforms (if there are any unwanted consequences). I haven't used Git & Github before so can anyone give me some directions please ... should I create pull request and continue discussion there?

**Comment by Jjagg** [comment]
@zigzag312 If you join the Gitter chat room (https://gitter.im/MonoGame/MonoGame) and pm me I'll walk you through it tomorrow.

**Comment by yyr57695** [comment]
One thing I did not notice before: under XNA, my game uses anti-aliasing.  Under both 3.5.1 and the latest dev versions of MonoGame, no anti-aliasing is present at all.

> yyr57695 do you also have multisampling enabled? Multisampling could also explain performance drop.

Yes, I do.  But it was also enabled under XNA.

I disabled it and noticed an immediate performance boost under MG 3.5.1.  The framerate had already been decent, but now seemed more stable.  I didn't notice much difference in my visuals, which are nothing all that special to begin with.  Anti-aliasing was still not present.

I then installed the latest MG dev build.  If you recall, earlier dev builds had dropped my game's performance, and with the latest build, the game didn't even start; I got a NoSuitableGraphicsDeviceException instead.  But after turning off multisampling, I was surprised to find that not only does my game start under the latest MG dev build, it runs at the just-about-60fps framerate I had before!

If I try to turn multisampling back on, I once again get the NoSuitableGraphicsDeviceException.

Anti-aliasing is still not present.  I'm really glad about the performance, but I would also like my anti-aliasing back at some point!

Hopefully this helps.  Any input would be appreciated, as well.

**Comment by zigzag312** [comment]
I think that anti-aliasing not working in 3.5.1 is known issue. Earlier dev builds versions of MonoGame tried to fix it, so if anti-aliasing was actually working in earlier dev build you tried, that would explain performance drop when going from 3.5.1 (where anti-aliasing wasn't working). 

NoSuitableGraphicsDeviceException is caused by some changes connected with anti-aliasing in current dev version. #5477 should fix that error when merged into develop branch.

**Comment by KonajuGames** [comment]
There is a fix for multisampling going in soon. See #5477 for more information.

**Comment by KonajuGames** [comment]
Ahh, didn't see that @zigzag312 already linked to that PR.

**Comment by Jjagg** [comment]
The multisample issue was fixed in #5477. No one else has reported a similar performance issue. I think it might be driver related. Closing this for now.
